crates/cext/src/circuit.rs

```diff
@@ -378,6 +381,59 @@ pub struct OpCounts {
     len: usize,
 }
 
+/// @ingroup QkCircuit
+/// Append an arbitrary unitary matrix to the circuit.
+/// The user passes a row-major array of interleaved real-imaginary pairs #[
+/// representing the unitary matrix.
+///
+/// @param curcuit      A pointer to the circuit to add the unitary to.
+/// @param num_qubits   The width of unitary gate.
+/// @param qubits       The pointer to array of quibit indices.
+/// The array length must be equal to num_qubits.
+/// @param matrix       The pointer to the row-major `(re, im)` doucle array///
+/// # Example
+///
+///
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_unitary(
+    circuit: *mut CircuitData,
+    num_qubits: u32,
+    qubits: *const u32,
+    matrix: *const f64,
+) -> ExitCode {
+    // SAFETY: Caller quarantees pointer validation, alignment
+    let circuit = unsafe { mut_ptr_as_ref(circuit) };
+
+    // Dimension of the unitart: 2^n
+    let dim = 1 << num_qubits;
+
+    // Copy (re, im) pairs to Complex64
+    let raw = unsafe { std::slice::from_raw_parts(matrix, dim * dim * 2) };
+    let data: Vec<Complex64> = raw
+        .chunks_exact(2)
+        .map(|x| Complex64::new(x[0], x[1]))
+        .collect();
+
+    // Build ndarray::Array2
+    let mat = Array2::from_shape_vec((dim, dim), data).expect("Invalid shape for unitary matrix");
```

Comment:
I wonder if using [`ArrayView::from_shape_ptr()`](https://docs.rs/ndarray/latest/ndarray/type.ArrayView.html#method.from_shape_ptr) here to create a view of the C array and then doing [`to_owned()`](https://docs.rs/ndarray/latest/ndarray/type.ArrayView.html#method.to_owned) to create a copy of it. I guess the advantage of doing the `ArrayView` path is that it lets you control the striding too and support various memory layouts. While doing it like this explicit memory layout for what we store in Rust for the `UnitaryGate`. TBH, I'm fine either way as long as we document it, but I wanted to raise it as a question to discuss at least.